engine/consensus/message_hub/message_hub.go

```diff
@@ -0,0 +1,517 @@
+	err := h.con.Unicast(packed.vote, packed.recipientID)
+	if err != nil {
+		log.Err(err).Msg("could not send vote")
+		return nil
+	}
+	// TODO(active-pacemaker): update metrics
+	//h.engineMetrics.MessageSent(metrics.EngineCompliance, metrics.MessageBlockVote)
+	log.Info().Msg("block vote transmitted")
+
+	return nil
+}
+
+// processQueuedProposal performs actual processing of model.Proposal, as a result of successful invocation
+// broadcasts block proposal to consensus committee and rest of Flow network.
+// No errors are expected during normal operations.
+func (h *MessageHub) processQueuedProposal(header *flow.Header) error {
+	// first, check that we are the proposer of the block
+	if header.ProposerID != h.me.NodeID() {
+		return fmt.Errorf("cannot broadcast proposal with non-local proposer (%x)", header.ProposerID)
+	}
+
+	// get the parent of the block
+	parent, err := h.headers.ByBlockID(header.ParentID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve proposal parent: %w", err)
+	}
+
+	// fill in the fields that can't be populated by HotStuff
+	header.ChainID = parent.ChainID
+	header.Height = parent.Height + 1
+
+	// retrieve the payload for the block
+	payload, err := h.payloads.ByBlockID(header.ID())
+	if err != nil {
+		return fmt.Errorf("could not retrieve payload for proposal: %w", err)
+	}
+
+	log := h.log.With().
+		Str("chain_id", header.ChainID.String()).
+		Uint64("block_height", header.Height).
+		Uint64("block_view", header.View).
+		Hex("block_id", logging.Entity(header)).
+		Hex("parent_id", header.ParentID[:]).
+		Hex("payload_hash", header.PayloadHash[:]).
+		Int("guarantees_count", len(payload.Guarantees)).
+		Int("seals_count", len(payload.Seals)).
+		Int("receipts_count", len(payload.Receipts)).
+		Time("timestamp", header.Timestamp).
+		Hex("signers", header.ParentVoterIndices).
+		//Dur("delay", delay).
+		Logger()
+
+	log.Debug().Msg("processing proposal broadcast request from hotstuff")
+
+	// TODO(active-pacemaker): replace with pub/sub?
+	h.hotstuff.SubmitProposal(header, parent.View) // non-blocking
+
+	// Retrieve all consensus nodes (excluding myself).
+	// CAUTION: We must include also nodes with weight zero, because otherwise
+	//          new consensus nodes for the next epoch are left out.
+	// Note: retrieving the final state requires a time-intensive database read.
+	//       Therefore, we execute this in a separate routine, because
+	//       `BroadcastTimeout` is directly called by the consensus core logic.
+	allIdentities, err := h.state.AtBlockID(header.ParentID).Identities(filter.And(
+		filter.Not(filter.HasNodeID(h.me.NodeID())),
+	))
+	if err != nil {
+		return fmt.Errorf("could not get identities for broadcasting proposal: %w", err)
+	}
+
+	recipients := allIdentities.Filter(filter.HasRole(flow.RoleConsensus))
+
+	// NOTE: some fields are not needed for the message
+	// - proposer ID is conveyed over the network message
+	// - the payload hash is deduced from the payload
```

Comment:
I'm not sure this comment is still relevant, especially after incorporating https://github.com/onflow/flow-go/pull/3372.

These fields will be included in the header which is sent over the network.

(Likewise for collection message hub)

Reply:
Yup, the plan is to merge `MessageHub` first and then apply those changes in https://github.com/onflow/flow-go/pull/3372.